File: min_time_apples_tree.rs
use std::collections::HashSet;
struct Solution;

impl Solution {
    pub fn min_time(n: i32, edges: Vec<Vec<i32>>, has_apple: Vec<bool>) -> i32 {
        let mut adj_list: Vec<Vec<i32>> = vec![vec![]; n as usize];
        for e in edges {
            adj_list[e[0] as usize].push(e[1]);
            adj_list[e[1] as usize].push(e[0]);
        }
        let mut total = 0i32;
        let mut vis: HashSet<i32> = HashSet::new();
        Self::dfs(0, adj_list, &mut vis, has_apple, &mut total);
        total
    }
    fn dfs(node: i32, adj_list: Vec<Vec<i32>>, vis: &mut HashSet<i32>,
           has_apple: Vec<bool>, total: &mut i32) -> bool {
        let mut res = false;
        if has_apple[node as usize] {
            res = true;
        }
        vis.insert(node);
        for i in &adj_list[node as usize] {
            if !vis.contains(&i) {
                if Self::dfs(*i, adj_list, vis, has_apple, total) {
                    *total += 2;
                    res = true;
                }
            }
        }
        res
    }
}

fn main() {
    let edges = vec![
        vec![0,1], vec![0,2],
        vec![1,4], vec![1,5],
        vec![2,3], vec![2,6]
    ];
    let has_apple1 = vec![false,false,true,false,true,true,false];
    let has_apple2 = vec![false,false,true,false,false,true,false];
    let has_apple3 = vec![false,false,false,false,false,false,false];
    println!("{}", Solution::min_time(7, edges, has_apple1)); //expect: 8
    println!("{}", Solution::min_time(7, edges, has_apple2)); //expect: 6
    println!("{}", Solution::min_time(7, edges, has_apple3)); //expect: 0
}

File: min_time_apples_tree.cpp
#include "leetcode.h"

class Solution {
public:
  int minTime(int n, vvd & edges, vector<bool> &hasApple) {
    unordered_map<int, int> um;
    int ans = 0;
    sort(edges.begin(), edges.end());
    for (auto const &e : edges) {
      if (um.count(e[1]) > 0)
        um[e[0]] = e[1];
      else
        um[e[1]] = e[0];
    }
    for (int i = 0; i < hasApple.size(); ++i) {
      if (!hasApple[i])
        continue;
      int parent = i;
      while (parent != 0 && um[parent] >= 0) {
        auto tmp = um[parent];
        um[parent] = -1;
        parent = tmp;
        ans += 2;
      }
    }
    return ans;
  }
};

int main() {
  Solution obj;
  vvd edges = {{0, 1}, {0, 2}, {1, 4}, {1, 5}, {2, 3}, {2, 6}};
  vector<bool> hasApple1 = {false, false, true, false, true, true, false},
               hasApple2 = {false, false, true, false, false, true, false},
               hasApple3 = {false, false, false, false, false, false, false};
  cout << obj.minTime(7, edges, hasApple1) << endl; // expect: 8
  cout << obj.minTime(7, edges, hasApple2) << endl; // expect: 6
  cout << obj.minTime(7, edges, hasApple3) << endl; // expect: 0
}
